src/Type/Php/PregSplitDynamicReturnTypeExtension.php

```diff
@@ -36,17 +48,125 @@ public function isFunctionSupported(FunctionReflection $functionReflection): boo
 
 	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): ?Type
 	{
-		$flagsArg = $functionCall->getArgs()[3] ?? null;
+		$args = $functionCall->getArgs();
+		if (count($args) < 2) {
+			return null;
+		}
+		$patternArg = $args[0];
+		$subjectArg = $args[1];
+		$limitArg = $args[2] ?? null;
+		$flagArg = $args[3] ?? null;
+		$patternType = $scope->getType($patternArg->value);
+		$patternConstantTypes = $patternType->getConstantStrings();
+		$subjectType = $scope->getType($subjectArg->value);
+		$subjectConstantTypes = $subjectType->getConstantStrings();
+
+		if (
+			count($patternConstantTypes) > 0
+			&& @preg_match($patternConstantTypes[0]->getValue(), '') === false
+		) {
+			return new ErrorType();
+		}
+
+		if ($limitArg === null) {
+			$limits = [-1];
+		} else {
+			$limitType = $scope->getType($limitArg->value);
+			$limits = $limitType->getConstantScalarValues();
+		}
+
+		if ($flagArg === null) {
+			$flags = [0];
+		} else {
+			$flagType = $scope->getType($flagArg->value);
+			$flags = $flagType->getConstantScalarValues();
+		}
```

Comment:
By replacing it as follows, type checking within multiple Constant loops will no longer be necessary.

```php
$flags = [];
$flagType = $scope->getType($flagArg->value);
foreach ($flagType->getConstantScalarValues() as $flag) {
    if (!is_int()) {
        return new ErrorType();
    }

    $flags[] = $flag;
}
```

Reply:
resolved https://github.com/phpstan/phpstan-src/pull/3757/commits/8cb3030cbe2e0cbf4da1d42bb9198d450d136df9